File: server/src/main/kotlin/com/splitme/core/DebtSimplifier.kt
package com.splitme.core

import com.splitme.model.TransferRoute
import kotlin.math.abs
import kotlin.math.min

object DebtSimplifier {
    private const val EPSILON = 0.0001

    fun simplifyDebts(netBalances: Map<String, Double>): List<TransferRoute> {
        val debtors = mutableMapOf<String, Double>() // netBalance < 0 => debt > 0
        val creditors = mutableMapOf<String, Double>() // netBalance > 0 => credit > 0

        for ((memberId, balance) in netBalances) {
            if (balance < -EPSILON) {
                debtors[memberId] = -balance
            } else if (balance > EPSILON) {
                creditors[memberId] = balance
            }
        }

        val transfers = mutableListOf<TransferRoute>()

        // 1. Exact match optimization: find pairs where debt == credit
        val debtorKeys = debtors.keys.toList()
        for (dKey in debtorKeys) {
            val dAmount = debtors[dKey] ?: continue
            val matchedCreditor = creditors.entries.firstOrNull { abs(dAmount - it.value) < EPSILON }
            if (matchedCreditor != null) {
                transfers.add(
                    TransferRoute(
                        fromMemberId = dKey,
                        toMemberId = matchedCreditor.key,
                        amount = dAmount
                    )
                )
                debtors.remove(dKey)
                creditors.remove(matchedCreditor.key)
            }
        }

        // 2. Greedy matching with remaining debts and credits
        while (debtors.isNotEmpty() && creditors.isNotEmpty()) {
            val maxDebtor = debtors.maxByOrNull { it.value } ?: break
            val maxCreditor = creditors.maxByOrNull { it.value } ?: break

            val transferAmount = min(maxDebtor.value, maxCreditor.value)

            transfers.add(
                TransferRoute(
                    fromMemberId = maxDebtor.key,
                    toMemberId = maxCreditor.key,
                    amount = transferAmount
                )
            )

            val remainingDebt = maxDebtor.value - transferAmount
            val remainingCredit = maxCreditor.value - transferAmount

            if (remainingDebt > EPSILON) {
                debtors[maxDebtor.key] = remainingDebt
            } else {
                debtors.remove(maxDebtor.key)
            }

            if (remainingCredit > EPSILON) {
                creditors[maxCreditor.key] = remainingCredit
            } else {
                creditors.remove(maxCreditor.key)
            }
        }

        return transfers
    }
}

File: client/src/core/debtSimplifier.ts
import { TransferRoute } from '../types/models';

const EPSILON = 0.0001;

export function simplifyDebts(netBalances: Record<string, number>): TransferRoute[] {
  const debtors: Record<string, number> = {};
  const creditors: Record<string, number> = {};

  for (const [memberId, balance] of Object.entries(netBalances)) {
    if (balance < -EPSILON) {
      debtors[memberId] = -balance;
    } else if (balance > EPSILON) {
      creditors[memberId] = balance;
    }
  }

  const transfers: TransferRoute[] = [];

  // 1. Exact match optimization: find pairs where debt == credit
  const debtorKeys = Object.keys(debtors);
  for (const dKey of debtorKeys) {
    const dAmount = debtors[dKey];
    if (dAmount === undefined) continue;

    for (const [cKey, cAmount] of Object.entries(creditors)) {
      if (Math.abs(dAmount - cAmount) < EPSILON) {
        transfers.push({
          fromMemberId: dKey,
          toMemberId: cKey,
          amount: dAmount,
        });
        delete debtors[dKey];
        delete creditors[cKey];
        break;
      }
    }
  }

  // 2. Greedy matching with remaining debts and credits
  while (Object.keys(debtors).length > 0 && Object.keys(creditors).length > 0) {
    let maxDebtorKey = '';
    let maxDebt = -Infinity;
    for (const [dKey, dAmount] of Object.entries(debtors)) {
      if (dAmount > maxDebt) {
        maxDebt = dAmount;
        maxDebtorKey = dKey;
      }
    }

    let maxCreditorKey = '';
    let maxCredit = -Infinity;
    for (const [cKey, cAmount] of Object.entries(creditors)) {
      if (cAmount > maxCredit) {
        maxCredit = cAmount;
        maxCreditorKey = cKey;
      }
    }

    if (!maxDebtorKey || !maxCreditorKey) break;

    const transferAmount = Math.min(maxDebt, maxCredit);

    transfers.push({
      fromMemberId: maxDebtorKey,
      toMemberId: maxCreditorKey,
      amount: transferAmount,
    });

    const remainingDebt = maxDebt - transferAmount;
    const remainingCredit = maxCredit - transferAmount;

    if (remainingDebt > EPSILON) {
      debtors[maxDebtorKey] = remainingDebt;
    } else {
      delete debtors[maxDebtorKey];
    }

    if (remainingCredit > EPSILON) {
      creditors[maxCreditorKey] = remainingCredit;
    } else {
      delete creditors[maxCreditorKey];
    }
  }

  return transfers;
}
